Added mouse input support for level viewer

lpenguin opened the pull request:
* Added two methods to `Application`
```rust
fn on_mouse_wheel(&mut self, MouseScrollDelta);`
fn on_mouse_move(&mut self, delta_x: f32, delta_y: f32, alt: bool);
```
* added few local variables to `main_loop`
```rust
let mut last_mouse_x: f32 = -1.0;
let mut last_mouse_y: f32 = -1.0;
let mut mouse_pressed: bool = false;
let mut alt_pressed: bool = false;
```
* added input events and handlers
```rust
 DepQuant {dir: f32}, // height shift
 PlaneQuant {x: f32, y: f32}, // x/y shift
 RotQuant {x: f32, y: f32}, // camera rotation
```

kvark commented on bin/boilerplate.rs:
```diff
@@ -67,6 +69,10 @@ impl Harness {
         let mut encoder = gfx::Encoder::from(self.factory.create_command_buffer());
         let mut last_time = time::Instant::now();
         let mut running = true;
+        let mut last_mouse_x: f32 = -1.0;
```
let's make this a tuple?

kvark commented on bin/boilerplate.rs:
```diff
@@ -7,12 +7,14 @@ extern crate vangers;
     fn on_resize<F: gfx::Factory<R>>(&mut self, render::MainTargets<R>, &mut F);
     fn on_key(&mut self, KeyboardInput) -> bool;
+    fn on_mouse_wheel(&mut self, MouseScrollDelta);
+    fn on_mouse_move(&mut self, delta_x: f32, delta_y: f32, alt: bool);
```
similarly, let's have `delta` as a tuple

kvark replied:
we shouldn't put `alt` here, since it's not a mouse related thing. The client can always intercept Alt themselves

kvark commented on bin/level/app.rs:
```diff
@@ -165,6 +185,33 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+                self.input = Input::Empty;
+            }
+            Input::PlaneQuant {x, y} => {
+                let mut vec_x = self.cam.rot * cgmath::Vector3::unit_x() * -1.0 * x;
```
can use unary `-` operator instead of `* - 1.0`, e.g. just ` * -x`

kvark commented on bin/level/app.rs:
```diff
@@ -165,6 +185,33 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+                let y1 = vec.y;
+                vec.z = 0.0;
+                let norm = vec.magnitude();
+                vec = cgmath::Vector3{x: x1 * norm1/norm, y: y1 * norm1/norm, z: 0.0};
```
there is a shorter form of the constructor: `cgmath::vec3(x,y,z)`

kvark commented on config/settings.toml:
```diff
@@ -1,4 +1,4 @@
-data_path = "/opt/gog/Vangers/game" #Linux (example)
```
might leave the old path commented out as well, for GoG people ;)

kvark commented on bin/boilerplate.rs:
```diff
@@ -93,6 +99,44 @@ impl Harness {
+                            }
+                        }
+                        glutin::WindowEvent::MouseInput {state, button, ..} => {
+                            if button == MouseButton::Left {
```
this is quite specific and tailored for the current use in `level` binary. I think we should move some of that logic inside the `level` app, and keep the boiler plate generic, i.e. just pass down both `state` and `button` to the `Application` directly, let it deal with that.

kvark commented on bin/boilerplate.rs:
```diff
@@ -93,6 +99,44 @@ impl Harness {
+                            app.on_mouse_wheel(delta)
+                        }
+                        glutin::WindowEvent::CursorMoved {position, ..} => {
+                            if mouse_pressed {
```
similarly, let's just pass it down to the application directly.
I imagine a use case we might want to get that would highlight parts of the level based on where the mouse is, even when no buttons are pressed

kvark commented on bin/boilerplate.rs:
```diff
@@ -7,12 +7,14 @@ extern crate vangers;
 pub trait Application<R: gfx::Resources> {
     fn on_resize<F: gfx::Factory<R>>(&mut self, render::MainTargets<R>, &mut F);
     fn on_key(&mut self, KeyboardInput) -> bool;
+    fn on_mouse_wheel(&mut self, MouseScrollDelta);
```
let's have default implementations that do nothing for those mouse events, given that only one of the apps need them atm

lpenguin commented:
Thanks for comments! I refactored code as follows:
* Only raw event logic for boilerplate
```rust
    fn on_mouse_wheel(&mut self, _delta: MouseScrollDelta) {}
    fn on_mouse_move(&mut self, _position: (f64, f64)) {}
    fn on_mouse_button(&mut self, _state: ElementState, _button: MouseButton) {}
```
* removed unnecessary Application function implementations for road/model/car bins.
* PlaneQuant, RotQuant using Vector2
* moved input processing to LevelView, added internal variables
```rust
    _last_mouse_pos: cgmath::Vector2<f32>,
    _alt_button_pressed: bool,
    _mouse_button_pressed: bool,
```
* small code improvements

kvark commented on bin/boilerplate.rs:
```diff
@@ -94,6 +97,15 @@ impl Harness {
+                            app.on_mouse_move(position)
+                        }
+                        glutin::WindowEvent::MouseInput {state, button, ..} => {
+                            app.on_mouse_button(state, button, )
```
nit: extra `,`

kvark commented on bin/level/app.rs:
```diff
@@ -1,14 +1,17 @@
     Hor { dir: f32, alt: bool },
     Ver { dir: f32, alt: bool },
     Dep { dir: f32, alt: bool },
+    DepQuant {dir: f32},
```
could just do `DepQuant(f32)` and same for the other of added variants, assuming you aren't going to add more fields any time soon
Note: I used the named `dir` and `alt` mostly because storing pure `bool` without a name is always ambiguous

kvark commented on bin/level/app.rs:
```diff
@@ -17,6 +20,10 @@ pub struct LevelView<R: gfx::Resources> {
     cam: space::Camera,
     input: Input,
+
+    _last_mouse_pos: cgmath::Vector2<f32>,
```
shouldn't use `_` prefix, since these are actually used (in Rust, it's only needed to silence compiler warnings on unused stuff)

lpenguin replied:
thought that private variables naming convention

kvark replied:
FYI, private is everything that is not qualified with `pub`

kvark commented on bin/level/app.rs:
```diff
@@ -95,6 +105,43 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+        }
+
+        let shift = position_vec - self._last_mouse_pos;
+        if !self._alt_button_pressed {
```
nit: could do functionally like `self.input = if self.alt_button_pressed {...} else {...};`

kvark commented on bin/level/app.rs:
```diff
@@ -165,6 +213,36 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+            Input::PlaneQuant {dir} => {
+                let cgmath::Vector2{x, y} = dir;
+
+                let vec_x = self.cam.rot * cgmath::Vector3::unit_x() * -x;
```
now I see what you want here, could look nicer with `cgmath::vec3(-x, 0.0, 0.0)`

kvark commented on bin/level/app.rs:
```diff
@@ -165,6 +213,36 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+                let y1 = vec.y;
+                vec.z = 0.0;
+                let norm = vec.magnitude();
+                vec = cgmath::vec3(x1 * norm1/norm, y1 * norm1/norm, 0.0);
```
could do something like `vec *= norm1/norm;`

kvark commented on bin/level/app.rs:
```diff
@@ -165,6 +213,36 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
+                self.input = Input::Empty;
+            }
+            Input::RotQuant {dir} => {
+                let cgmath::Vector2{x, y} = dir;
```
this doesn't help much here, you can either pattern match it earlier as `Input::RotQuant { dir: cgmath::Vector2 { x, y }} => {...}` or just use `dir.x` and `dir.y` here

lpenguin replied:
Yes, nested pattern matching! Didn't think about that. May be it's too verbose, i think i'll just switch to `dir.x` in this case.

lpenguin commented:
@kvark thanks for comments again! i think i fixed everything. I'll squash all commits into one later
